marimo/_plugins/ui/_impl/chat/chat.py

```diff
@@ -0,0 +1,175 @@
+from __future__ import annotations
+
+from dataclasses import dataclass
+from typing import Any, Callable, Dict, Final, List, Optional
+
+from marimo._output.formatting import as_html
+from marimo._output.rich_help import mddoc
+from marimo._plugins.ui._core.ui_element import UIElement
+from marimo._plugins.ui._impl.chat.convert import model_from_callable
+from marimo._plugins.ui._impl.chat.types import (
+    ChatMessage,
+    ChatModel,
+    ChatModelConfig,
+    from_chat_message_dict,
+)
+from marimo._runtime.functions import EmptyArgs, Function
+
+
+@dataclass
+class SendMessageRequest:
+    messages: List[ChatMessage]
+    config: ChatModelConfig
+
+
+@dataclass
+class GetChatHistoryResponse:
+    messages: List[ChatMessage]
+
+
+DEFAULT_SYSTEM_MESSAGE = (
+    "You are a helpful assistant specializing in data science."
+)
+
+
+@mddoc
+class chat(UIElement[Dict[str, Any], List[ChatMessage]]):
+    """
+    A chatbot UI element for interactive conversations.
+
+    **Example - Using a custom model.**
+
+    You can define a custom chat model that takes a
+    prompt, previous messages, and optional attachments as input
+    and returns a response.
+
+    The response can be an object, a marimo UI element, or a string.
+
+    ```python
+    def my_rag_model(prompt, messages, attachments):
+        docs = find_docs(prompt)
+        prompt = template(prompt, docs, messages)
+        response = query(prompt)
+        if is_dataset(response):
+            return dataset_to_chart(response)
+        return response
+
+
+    chat = mo.ai.chatbot(
+        my_rag_model,
+        system_message="You are a helpful assistant.",
+    )
+    ```
+
+    **Example - Using a built-in model.**
+
+    You can use a built-in model from the `mo.ai` module.
+
+    ```python
+    chat = mo.ai.chatbot(
+        mo.ai.openai("gpt-4o"),
+        system_message="You are a helpful assistant.",
+    )
+    ```
+
+    **Attributes.**
+
+    - `value`: the current chat history
+
+    **Initialization Args.**
+
+    - `model`: the chatbot model
+    - `system_message`: the initial system message for the chatbot
+    - `prompts`: optional list of prompts to start the conversation
+    - `on_message`: optional callback function to handle new messages
+    - `max_tokens`: maximum number of tokens in the response
+    - `temperature`: sampling temperature for response generation
+    """
+
+    _name: Final[str] = "marimo-chatbot"
+
+    def __init__(
+        self,
+        model: ChatModel,
```

Comment:
This can also be a callable, right?

I think I'm finding it a bit confusing having the callable API mixed with the canned model API. Since some parameters it seems (?) are only relevant for when you're using a canned model.

Should the public API be split into two APIs?

`mo.ui.chat.from_model(...)`

`mo.ui.chat.from_callable(...)`

?

Reply:
I prefer the composition - I think it's a cleaner api

Reply:
I can make the canned APIs callable as well - I just find oop cleaner and that's all internal anyways